docs/linux/setup_linux-host_qemu-vm_riscv64-kernel.md

```diff
@@ -0,0 +1,153 @@
+patch. Create a kernel config with:
+
+```shell
+make ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- defconfig
+```
+
+In addition enable the [recommended Kconfig options for syzkaller](/docs/linux/kernel_configs.md).
+
+Then build kernel with:
+
+```
+make ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- -j $(nproc) vmlinux
+```
+
+# BBL (Berkeley Boot Loader)
+
+Clone the BBL repository and build the bootable image containg bootloader and kernel:
+
+```shell
+git clone https://github.com/riscv/riscv-pk
+cd riscv-pk
+mkdir build && cd build
+../configure --enable-logo --host=riscv64-unknown-elf --with-payload=/path/to/linux/vmlinux
+make
+```
+
+# Image
+
+We will use buildroot to create the disk image. You can obtain buildroot
+[here](https://buildroot.uclibc.org/download.html). The following instructions
+were tested with buildroot version 2020.05. First run:
+
+```shell
+make qemu_riscv64_virt_defconfig
+make menuconfig
+```
+
+Choose the following options:
+
+```
+    Target packages
+	    Networking applications
+	        [*] dhcpcd
+	        [*] iproute2
+	        [*] openssh
+    Filesystem images
+                ext2/3/4 variant - ext4
+	        exact size - 1g
+```
+
+Unselect:
+
+```
+    Kernel
+	    Linux Kernel
+```
+
+Run `make`.
+
+Then add the following line to `output/target/etc/fstab`:
+
+```
+debugfs	/sys/kernel/debug	debugfs	defaults	0	0
+```
+
+Then replace `output/target/etc/ssh/sshd_config` with the following contents:
+
+```
+PermitRootLogin yes
+PasswordAuthentication yes
+PermitEmptyPasswords yes
+```
+
+Run `make` again.
+
+# Test kernel and image
+
+Run:
+
+```shell
+qemu-system-riscv64 \
+	-machine virt \
+	-nographic \
+	-kernel /riscv-pk/build/bbl \
+	-append "root=/dev/vda ro console=ttyS0" \
+	-object rng-random,filename=/dev/urandom,id=rng0 \
+	-device virtio-rng-device,rng=rng0 \
+	-drive file=/buildroot/output/images/rootfs.ext2,if=none,format=raw,id=hd0 \
+	-device virtio-blk-device,drive=hd0 \
+	-netdev user,id=net0,host=10.0.2.10,hostfwd=tcp::10022-:22 \
+	-device virtio-net-device,netdev=net0
+```
+
+This should boot the kernel. Wait for login prompt, then in another console run:
+
+```
+ssh -p 10022 root@localhost
+```
+
+ssh should succeed.
+
+# syzkaller
+
+Build syzkaller as described [here](/docs/linux/setup.md#go-and-syzkaller), with `riscv64` target:
+
+```
+make TARGETOS=linux TARGETARCH=riscv64
+```
+
+Create the manager config `riscv64.cfg` similar to the following one (adjusting paths as necessary):
+
+```
+{
+	"name": "riscv64",
+	"target": "linux/riscv64",
+	"http": ":56700",
+	"workdir": "/workdir",
+	"kernel_obj": "/linux",
+	"syzkaller": "/gopath/src/github.com/google/syzkaller",
+	"image": "/buildroot/output/images/rootfs.ext2",
+	"procs": 8,
+	"type": "qemu",
+	"vm": {
+		"count": 1,
+		"qemu_args": "-machine virt -device virtio-blk-device,drive=hd0",
+		"image_device": "drive if=none,format=raw,id=hd0,file=",
+		"net_device_type": "virtio-net-device",
```

Comment:
Can we do without adding this config? Are these required? Or optional?

The idea is that qemu VM should work without specifying any additional parameters (qemu_args, image_device, net_device_type) in some "recommended" configuration. Advanced users may override something via qemu_args. But otherwise we try to not require lots of knowledge re all these qemu flags (frankly I always lost in all these flags and how they interact, qemu man usually don't answer any questions).
ARM has zoo of configurations, but hopefully for risc-v at this early stage we can do "it just works out-of-the-box".

For example for "virtio-net-device" we probably could switch all arches to this new way? I can imagine it can break things. So maybe we can start by adding "this arch uses new way of setting up network" flag to archConfig, and then just pass the necessary flags without asking user.
For image flags, it would also be good to somehow fit it into archConfig.

Reply:
> Can we do without adding this config? Are these required? Or optional?

In the current form of the PR, these are required which I agree is not optimal.

> The idea is that qemu VM should work without specifying any additional parameters (qemu_args, image_device, net_device_type) in some "recommended" configuration. Advanced users may override something via qemu_args. But otherwise we try to not require lots of knowledge re all these qemu flags (frankly I always lost in all these flags and how they interact, qemu man usually don't answer any questions).
> ARM has zoo of configurations, but hopefully for risc-v at this early stage we can do "it just works out-of-the-box".

Yeah, I got confused by the myriad of options as well. https://www.qemu.org/2018/05/31/nic-parameter/ helped a bit to untangle the confusion around the network options for me. I was thinking about switching to the "new options" (i.e. `-device ... -netdev`) in general but was worried it might break existing ports.

> For example for "virtio-net-device" we probably could switch all arches to this new way? I can imagine it can break things. So maybe we can start by adding "this arch uses new way of setting up network" flag to archConfig, and then just pass the necessary flags without asking user.
> For image flags, it would also be good to somehow fit it into archConfig.

Good point. I'll try to respin the PR to make the "new options" the default for new ports (i.e. riscv64 at the moment) and drop the `net_device_type` config. Ideally the VM config should then look like:

```
        "vm": {
                "count": 1,
                "kernel": "...",
                "cpu": 2,
                "mem": 2048
        }
```